blueprints/traffic-ops-cache-config-endpoint.md

```diff
@@ -0,0 +1,155 @@
+
+The data currently needed by the cache config generation which will be snapshotted can be seen here:
+
+https://godoc.org/github.com/apache/trafficcontrol/traffic_ops/ort/atstccfg/config#TOData
+
+### Traffic Portal Impact
+
+n/a
+
+### Traffic Ops Impact
+
+As above, TO should cache the snapshot in memory, and only fetch from the database if the cache has expired, and pre-check if the timestamp is unchanged (common) before initiating the full expensive snapshot retrieval.
+
+- The snapshot should use DeliveryService-Server IDs, not names, and omit the DSS last_modified, for both the database query and the response.
+    - DSS is 99% of the snapshot, using IDs and omitting the timestamp is a large size  reduction, saving network both DB->TO  and TO->client.
+    - The snapshot should use one-letter JSON names for DSS, i.e. `“d”: 42, “s”: 24`. This will further reduce the network to the client by about 50%, and improve TO scalability.
+    - The snapshot cache should store the serialized JSON bytes, so it doesn’t have to re-serialize for every request. JSON serializing in Go is a huge performance cost, and this will drastically improve scalability.
+
+Queueing Updates will automatically create the Snapshot
+    - This may be changed to a button on the Server in the future. But for now, automatically snapshotting on Queue avoids changing the User Interface.
+
+GET `/servers/{{name}}/cache-config-snapshot`
+    - Returns the cache config snapshot
+    - Per-server, because a lot of data is only needed by one server. For example, that server’s Parameters.
+    - TO should load the Snapshot, and then remove the unnecessary data before serializing it in the GET for one server.
+
+#### REST API Impact
+
+See Traffic Ops Impact
+
+#### Client Impact
+
+See Traffic Ops Impact - client functions for each endpoint.
+
+#### Data Model / Database Impact
+
+New table for Cache Config Snapshots.
+
+### ORT Impact
+
+ORT will be changed to request this endpoint if it exists, instead of the API endpoints.
+
+This should be a small change, and should only require changing the single request function. See https://github.com/apache/trafficcontrol/blob/b3e8605e9bdcaf097372c9c5dff337b72ff0bc66/traffic_ops/ort/atstccfg/cfgfile/cfgfile.go#L43
+
+ORT will need to continue to request the API endpoints for the prior Traffic Ops version, if the endpoint doesn't exist, for at least one major Traffic Control version.
+
+### Traffic Monitor Impact
+
+n/a
+
+### Traffic Router Impact
+
+n/a
+
+### Traffic Stats Impact
+
+n/a
+
+### Traffic Vault Impact
+
+n/a
+
+### Documentation Impact
+
+The new API endpoint will be documented. See Traffic Ops Impact.
+
+### Testing Impact
+
+The new API endpoint, and ATS config changes, will have TO API tests and unit tests. See Traffic Ops Impact.
+
+### Performance Impact
+
+No significant performance impact expected. Load impact on Traffic Ops should be reduced.
+
+### Security Impact
+
+n/a
```

Comment:
We may want to note here that secrets required by ORT (sslkeys, URI signing keys, etc) would now be stored in the cache-config-snapshot JSON blob in TODB, rather than existing only in Riak, so this new endpoint should require admin privileges.